### PR TITLE
KAFKA-10199: Cleanup TaskManager and Task interfaces

### PR DESCRIPTION
**Pull request by guozhangwang**
This PR is to consolidate https://github.com/apache/kafka/pull/12338 and https://github.com/apache/kafka/pull/12384

In order to integrate with the state updater, we would need to refactor the TaskManager and Task interfaces. This PR achieved the following purposes:

1. Separate active and standby tasks in the Tasks placeholder, plus adding pendingActiveTasks and pendingStandbyTasks into Tasks. The exposed active/standby tasks from the Tasks set would only be mutated by a single thread, and the pending tasks hold for those tasks that are assigned but cannot be actively managed yet. For now they include two scenarios: a) tasks from unknown sub-topologies and hence cannot be initialized, b) tasks that are pending for being recycled from active to standby and vice versa. Note case b) would be added in a follow-up PR.

2. Extract any logic that mutates a task out of the Tasks / TaskCreators. Tasks should only be a place for maintaining the set of tasks, but not for manipulations of a task; and TaskCreators should only be used for creating the tasks, but not for anything else. These logic are all migrated into TaskManger.

3. While doing 2) I noticed we have a couple of minor issues in the code where we duplicate the closing logics, so I also cleaned them up in the following way:
a) When closing a task, we first trigger the corresponding closeClean/Dirty function; then we remove the task from `Tasks` bookkeeping, and for active task we also remove its task producer if EOS-V1 is used.
b) For closing dirty, we swallow the exception from close call and the remove task producer call; for closing clean, we store the thrown exception from either close call or the remove task producer, and then rethrow at the end of the caller. The difference though is that, for the exception from close call we need to retry close it dirty; for the exception from the remove task producer we do not need to re-close it dirty.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdaterTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdaterTest.java]: This is a leftover from previous commit, we do not need this anymore.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java:64]: This is mainly for case 1) in the description.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java:145]: This is for case 2) in description.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTaskCreator.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTaskCreator.java:76]: This is for case 2) in description.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:295]: Minor cleanup: duplicated logging.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:582]: Consolidate the logic of recycling within a task, so that it can be extracted from the creator later for recycling after tasks returned from state updater.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:383]: This is a cleanup: we construct the pending tasks to recycle outside the handleCloseAndRecycle so that we do not need to pass in activeTasksToCreate / standbyTasksToCreate anymore.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:862]: This is for case 3) in the description: this is already in the `completeTaskCloseClean` so is a duplicated code.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:111]: Minor cleanup: do not allow callers to modify the set.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:156]: Minor cleanup: do deep copy since the passed in `topicPartition` param maybe modified.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:302]: This is for case 2) in description: extract all `task.mutate` functions out of `Tasks` and only inside `TaskManager`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java:68]: This is for case 2) in description: we detect unknown named topologies in the caller (`TaskManager`) so `Tasks` would be just dumb book-keeping.

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java
Comment (cadonna) [inline clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: This seems a bit weird to me. I see that you changed this because `Tasks#allTaskIds()` returns a `Collection` instead of a `Set`. I think it is fine to let `Tasks#allTaskIds()` return a `Set`. The other reason you changed that is that is `Tasks#allTasks()` computing its result from two collections. Those two collections can be transformed to sets. Maybe we should also consider to change the return type of `allTasks()` to a set.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java]: I would not use a dedicated field for the task timeout but just directly use the task config object `config`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java]:
```suggestion
    // 2. They are to be recycled from an existing restoring task yet to be returned from the state updater.
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java]:
```suggestion
            throw new IllegalStateException("Attempted to remove a task that is not closed: " + taskId);
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:246]: Could you add unit tests for this method?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:582]: Could you add unit tests for this method?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:1055]: Could you please add unit tests for this method? I think you can reuse the one that was deleted in `TasksTest`.

**Review by cadonna — COMMENTED** [review]
Thanks for the PR, @guozhangwang ! 

Here my feedback.

I will try to base my PR on this one.

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: Good point, I also feel using `Set` is safer for de-dup effects. Will change it back.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java]: Ack.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java]: Oops, my bad :) thanks!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:246]: I re-think about this a bit, and I felt just having `recycle` in the task is not sufficient as a refactoring, since we'd need to still call 1) task.closeCleanAndRecycleState, 2) taskCreator.createXTaskFromY(), which would call 3) task.recycle() in order.

I think it's better to extract the creators out of the `Tasks` (then also extract the main consumer as well since it's only needed for the creation of the tasks) into `TaskManager`, and merge the above steps into a single function. But since this PR is getting complicated enough I'd probably defer this to another PR.

I could either revert this function or just keep it here without the test coverage, and then further consolidate the logic in a follow-up PR, which one do you prefer?

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:246]: Yeah, I was also wondering if we can simplify this chain of calls from `TaskManager` -> `Tasks` -> `StandbyTaskCreator`/`ActiveTaskCreator` -> `StreamTask`/`StandbyTask`. And I also thought it might be better to postpone this refactoring. 🙂

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java:219]: @cadonna This is a bug exposed from the unit test now: we should not modify the passed in params directly.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java:316]: I would even make it unmodifiable to make it clear that changing the returned map will not change the maps within the `Tasks` object.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java:313]: nit: `ret` is not really a meaningful variable name.

**Review by cadonna — APPROVED** [review]
LGTM!

I had minor comments which we can fix in a follow-up.

I waiting for an acceptable build before I merge.

**Comment by cadonna** [comment]
Build failures are unrelated.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java:316]: Sounds good

**Pull request merged**
